d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadLoadBalancer.java

```diff
@@ -109,35 +111,31 @@ public void getClient(Request request, RequestContext requestContext, Callback<T
         _newLb.getClient(request, requestContext, clientCallback);
         break;
       case DUAL_READ:
-        _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>()
-        {
-          @Override
-          public void onError(Throwable e)
-          {
-            LOG.error("Double read failure. Unable to read service properties from: " + serviceName, e);
-          }
-
-          @Override
-          public void onSuccess(ServiceProperties result)
-          {
-            String clusterName = result.getClusterName();
-            _dualReadStateManager.updateCluster(clusterName, DualReadModeProvider.DualReadMode.DUAL_READ);
-            _newLb.getLoadBalancedClusterAndUriProperties(clusterName, new Callback<Pair<ClusterProperties, UriProperties>>()
-            {
+        getLoadBalancerThreadPool().execute(
+            () -> _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>() {
               @Override
-              public void onError(Throwable e)
-              {
-                LOG.error("Dual read failure. Unable to read cluster properties from: " + clusterName, e);
+              public void onError(Throwable e) {
+                LOG.error("Double read failure. Unable to read service properties from: " + serviceName, e);
```

Comment:
Since you're modifying this code, can you fix this string concatenation in the logger?
```java
LOG.error("Double read failure. Unable to read service properties from: {}", serviceName, e);
```

Reply:
Upated~